Fixed issue with formatting the date in the movie list view HTML template.

kjcioffi opened the pull request:
- Implemented a method in movie api utils to handle converting date strings to datetime objects.
- Incorporated the method into the MovieListView.
- Added test coverage to support the feature.

kjcioffi committed:
Fixed issue with formatting the date in the movie list view HTML template.
   - Implemented a method in movie api utils to handle conversion of date strings to datetime objects.
   - Incorporated the method into the MovieListView.
   - Added test coverage to support the feature.

ryaustin commented on movies/movie_api_utils.py:
```diff
@@ -85,3 +85,23 @@ def _get(self, url, **kwargs) -> dict[str, Any]:
         except requests.exceptions.RequestException:
             raise MovieApiException("An error occurred while processing your request.")
+
+    def convert_date_string_into_object(self, movies, *, filter: str = None):
```
The docstring here could benefit from an example of how the filter would be used.
Is it to search for a certain movie? date?

ryaustin commented on movies/movie_api_utils.py:
```diff
@@ -85,3 +85,23 @@ def _get(self, url, **kwargs) -> dict[str, Any]:
+        Iterates through a collection of movies and converts their release date string into a datetime object.
+        """
+        try:
+            movies_to_process = movies if filter is None else movies[filter]
```
here is also a potential issue (if searching for a string pattern). You get a single result vs every result that matches your filter.

ryaustin replied:
Okay, looks like this is resolved.

ryaustin commented on movies/tests/test_movie_api_utils.py:
```diff
@@ -30,3 +30,20 @@ def test_movies_now_playing_success(self):
+        with self.assertRaisesMessage(
+            MovieApiException, f"{filter} is not a valid attribute on a movie object."
+        ):
+            self.movie_api_utils.convert_date_string_into_object(movies, filter=filter)
```
Include a test for a good filter.
- happy-path tests prove that what you intended works and allows for other developers to quickly understand the code by simply reviewing the tests.

ryaustin reviewed (changes requested):
Would like to get a better understanding on the filter mechanism.

ryaustin commented:
- [ ] take a look at the dateparser library apart of stdlib: https://dateparser.readthedocs.io/en/latest/

Dateparser handles errors and exceptions better than the datetime library and returns a datetime object. You cannot trust external data and one malformed date or null value can cause exceptions on your main page - this won't happen with dateparser.

You can consider using it in several ways. 
1. As a fill-in for .strftime as used in your code, 
2. As a custom template filter that you pass string dates (from the template).

kjcioffi committed:
- Added extra documentation around the filter keyword argument in convert_date_string_into_object method in MovieApiUtils
- Additional test coverage was added for checking against successful date conversions.

kjcioffi commented:
I was not aware date parser existed. I am most certainly interested in using it, however, I don't follow what you mean by using it as a custom template filter. 

Are we talking about something that can be used with an HTML template or is this still about the **filter** keyword argument?

ryaustin commented:
Awesome. date parser is a game changer when you discover it. You can almost through any kind of date at it and it will either convert it or fail gracefully.

Its outside of the scope of the project since you already created the method to convert strings (I believe this approach is best), but the custom filter would work in the same way the built-in date filter works as in: ` movie.release_date | date:"..."` 

It allows you to extend the template with functions that you can pass arguments to. So if you were to create a custom filter called string_to_date, your code becomes `movie.release_date | string_to_date:movie.release_date` or something like that. Checkout this link to learn more about this powerful pattern: https://docs.djangoproject.com/en/5.0/howto/custom-template-tags/

ryaustin reviewed (approved):
- filter has been more clearly defined
- happy path test gives an example of what is expected
- No regression expected
- Looks good to merge
